src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -24,39 +24,85 @@
 
 package net.fabricmc.loom.providers;
 
+import java.io.BufferedReader;
 import java.io.File;
 import java.io.IOException;
+import java.net.URL;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 import java.util.function.Consumer;
 
+import javax.annotation.Nullable;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.tools.ant.util.StringUtils;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Dependency;
 
 import net.fabricmc.loom.LoomGradleExtension;
 import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.DependencyProvider;
 import net.fabricmc.loom.util.Version;
-import net.fabricmc.mappings.Mappings;
+import net.fabricmc.mapping.tree.TinyMappingFactory;
+import net.fabricmc.mapping.tree.TinyTree;
+import net.fabricmc.stitch.Command;
 import net.fabricmc.stitch.commands.CommandProposeFieldNames;
+import net.fabricmc.stitch.commands.tinyv2.CommandMergeTinyV2;
+import net.fabricmc.stitch.commands.tinyv2.CommandReorderTinyV2;
 
-//TODO fix local mappings
-//TODO possibly use maven for mappings, can fix above at the same time
 public class MappingsProvider extends DependencyProvider {
 	public MinecraftMappedProvider mappedProvider;
 
 	public String mappingsName;
 	public String minecraftVersion;
 	public String mappingsVersion;
 
-	public File MAPPINGS_DIR;
-	public File MAPPINGS_TINY_BASE;
-	public File MAPPINGS_TINY;
-	public File MAPPINGS_MIXIN_EXPORT;
+	private boolean isV2;
+	private Path mappingsDir;
+	private Path mappingsStepsDir;
+	private Path tinyMappingsWithoutEnums;
+	public File tinyMappings;
+	public File mappingsMixinExport;
+
+	public void clean() throws IOException {
+		tinyMappings.delete();
+		FileUtils.deleteDirectory(mappingsStepsDir.toFile());
+	}
+
+	public TinyTree getMappings() throws IOException {
+		return MappingsCache.INSTANCE.get(tinyMappings.toPath());
+	}
+
+	@Nullable
+	public TinyTree getMappingsOfVersion(Project project, String version) throws IOException {
+		Path migrateMappingsDir = mappingsDir.resolve("migrate");
+		Path localMappingsOfVersion = migrateMappingsDir.resolve(version + ".tiny");
 
-	public Mappings getMappings() throws IOException {
-		return MappingsCache.INSTANCE.get(MAPPINGS_TINY.toPath());
+		if (Files.exists(localMappingsOfVersion)) {
+			try (BufferedReader reader = Files.newBufferedReader(localMappingsOfVersion)) {
+				return TinyMappingFactory.loadWithDetection(reader);
+			}
+		} else {
+			//TODO: download unmerged v2 mappings once they are on the maven (v2-yarn)
+			String versionRelativePath = version.replace("+", "%2B").replace(" ", "%20");
+			String artifactUrl = "https://maven.fabricmc.net/net/fabricmc/yarn/" + versionRelativePath + "/yarn-" + versionRelativePath + ".jar";
```

Comment:
~~Humm, why like this and not as a normal maven dep that the dev can change? This wont handle maven local, or anything else?~~

Seems to only be used by MigrateMappings would it fit better there? or even just lower down this class?

Reply:
IIrc it uses private MigrateMappings API

Reply:
It uses `mappingsDir` which I would rather not expose to outsiders. what do you think?